### PR TITLE
feat: add delete WAL in drop_region

### LINKED ISSUE #1589: Obsolete WAL entries when dropping table
### What type of bug is this?

Other

### What subsystems are affected?

Other

### What happened?

Currently GreptimeDB does not support `DROP TABLE` clause #497 . Before this issue is resolved, we need to find a way to clean the data in WAL so that WAL can push the obsolete offset forward so that disk space can be  reclaimed.

### What operating system did you use?

NA

### Relevant log output and stack trace

```bash
NA
```

### How can we reproduce the bug?

NA

### PR DESCRIPTION
**Pull request by DevilExileSu**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

add delete WAL in drop_region

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
fixes #1589
#1556

### CODE REVIEW

**Comment by evenyag** [comment]
Thanks, @DevilExileSu . I will review this before this Friday.

**Comment by DevilExileSu** [comment]
> Thanks, @DevilExileSu . I will review this before this Friday.

Thanks, actually I’m still a bit confused😂:
+ Before calling `delete_namespace()`  should I check if the namespace still exists, to prevent duplicate writing?
+ >  In [#1556](https://github.com/GreptimeTeam/greptimedb/issues/1556#issuecomment-1547172666).  They usually don't block the drop operation so we could remove the manifest before all files are deleted.  
     
    It seems that store the Remove meta action before deleting data in WAL is useless

**Comment by gitccl** [comment]
> > Thanks, @DevilExileSu . I will review this before this Friday.
> 
> Thanks, actually I’m still a bit confused😂:
> 
> * Before calling `delete_namespace()`  should I check if the namespace still exists, to prevent duplicate writing?
> * > In [#1556](https://github.com/GreptimeTeam/greptimedb/issues/1556#issuecomment-1547172666).  They usually don't block the drop operation so we could remove the manifest before all files are deleted.
>   
>   
>   It seems that store the Remove meta action before deleting data in WAL is useless

I think its necessary to remove meta action before deleteing data in WAL. Once the node is power off after we delete wal, the data is lost.

> > Thanks, @DevilExileSu . I will review this before this Friday.
> 
> Thanks, actually I’m still a bit confused😂:
> 
> * Before calling `delete_namespace()`  should I check if the namespace still exists, to prevent duplicate writing?
> * > In [#1556](https://github.com/GreptimeTeam/greptimedb/issues/1556#issuecomment-1547172666).  They usually don't block the drop operation so we could remove the manifest before all files are deleted.
>   
>   
>   It seems that store the Remove meta action before deleting data in WAL is useless

I think it is necessary to persist remove meta action before delete data in WAL in the current implementation. Otherwise the data will be lost once the node is power off immediately after wal is deleted.

**Comment by DevilExileSu** [comment]
> I think it is necessary to persist remove meta action before delete data in WAL in the current implementation. Otherwise the data will be lost once the node is power off immediately after wal is deleted.

I think it is possible to resume the previous drop_region operation by using Procedure after a power outage and reboot.

But another key issue is that the [on_engine_drop_table](https://github.com/GreptimeTeam/greptimedb/blob/715e1a321fc21ae546042dfd65d6b211cce26752/src/table-procedure/src/drop.rs#LL51C60-L51C60) operation is performed after [on_remove_from_catalog](https://github.com/GreptimeTeam/greptimedb/blob/715e1a321fc21ae546042dfd65d6b211cce26752/src/table-procedure/src/drop.rs#LL50C55-L50C77), after the table has been successfully removed from `SystemCatalogTable`. When executing `on_engine_drop_table`, some regions are deleted, but power outage happens. After rebooting, the corresponding table cannot be opened from `SystemCatalogTabel`. In this case, neither recover Procedure nor recover Manifest seems to be able to execute.

https://github.com/GreptimeTeam/greptimedb/blob/715e1a321fc21ae546042dfd65d6b211cce26752/src/mito/src/engine.rs#L615-L626

I add the following code, and simulate failback on `close()`
```rust
     if let Some((_, table)) = &removed_table { 
         table 
             .close() 
             .await 
             .map_err(BoxedError::new) 
             .context(table_error::TableOperationSnafu)?; 
  
         Ok(true) 
     } else { 
         debug!("fail to drop table. request = {:?}", request);
         Ok(false) 
     } 
 } 
```
Print the following log after reboot:
```
2023-05-17T04:20:02.624652Z DEBUG mito::engine: fail to drop table. request = DropTableRequest { catalog_name: "greptime", schema_name: "public", table_name: "monitor" }
2023-05-17T04:20:02.625122Z DEBUG common_procedure::local::runner: Execute procedure mito::DropMitoTable-fcf42257-543b-4071-a1e3-17e88b7265d0 once, status: Done, need_persist: false
2023-05-17T04:20:02.626066Z DEBUG write{path="procedure/fcf42257-543b-4071-a1e3-17e88b7265d0/0000000001.commit" args=OpWrite { content_length: Some(0), content_type: None, content_disposition: None, cache_control: None }}: opendal::services: service=fs operation=write path=procedure/fcf42257-543b-4071-a1e3-17e88b7265d0/0000000001.commit -> started
```

**Comment by gitccl** [comment]
> I think it is possible to resume the previous `drop_region` operation by using Procedure after a power off and reboot.

Oh, yeah. I forgot the procedure will also be persisted.

**Comment by evenyag** [comment]
We expect the storage engine can work without the procedure framework. It should maintain its internal state correctly. As you can see, both creating and altering a region follow the same routine.

We use procedures to handle failure when we want to modify multiple regions.

**Comment by DevilExileSu** [comment]
> We expect the storage engine can work without the procedure framework. It should maintain its internal state correctly. As you can see, both creating and altering a region follow the same routine.
> 
> We use procedures to handle failure when we want to modify multiple regions.

Got it! Thank you for your patience and reply.

File: src/storage/src/error.rs [hunk 1]
```diff
@@ -240,6 +240,17 @@ pub enum Error {
...
     },
 
+    #[snafu(display(
+        "Failed to delete WAL namespace, region id: {}, source: {}",
+        region_id,
+        source
+    ))]
+    DeleteWalNamespace {
+        region_id: RegionId,
+        #[snafu(backtrace)]
```
Comment (waynexia) [inline src/storage/src/error.rs:250]: This backtrace way is outdated, use location instead
```suggestion
        location: Location
```

File: src/storage/src/region.rs [hunk 2]
```diff
@@ -675,6 +680,20 @@ impl<S: LogStore> RegionInner<S> {
...
+        let drop_ctx = DropContext {
+            shared: &self.shared,
+            wal: &self.wal,
+            manifest: &self.manifest,
+            flush_scheduler: &self.flush_scheduler,
+            compaction_scheduler: &self.compaction_scheduler,
+            sst_layer: &self.sst_layer,
+        };
+        self.writer.on_drop(drop_ctx).await?;
+        self.manifest.stop().await
```
Comment (waynexia) [inline src/storage/src/region.rs:694]: Stop manifest before drop region looks better in my perspective

File: tests/cases/distributed/common
Comment (waynexia) [inline tests/cases/distributed/common:1]: Is this change intended or by accident

File: src/storage/src/sst.rs [hunk 3]
```diff
@@ -121,6 +121,16 @@ impl LevelMetas {
         merged
     }
 
+    pub fn mark_deleted_all_files(&self) -> Vec<String> {
```
Comment (waynexia) [inline src/storage/src/sst.rs:124]: nit:
```suggestion
    pub fn mark_all_files_deleted(&self) -> Vec<String> {
```

File: src/storage/src/region/writer.rs [hunk 4]
```diff
@@ -270,7 +271,58 @@ impl RegionWriter {
...
+        let prev_version = version_control.current_manifest_version();
+        action_list.set_prev_version(prev_version);
+
+        logging::debug!(
+            "Try to remove region {}, action_list: {:?}",
+            drop_ctx.shared.name(),
+            action_list
+        );
+
+        drop_ctx.manifest.update(action_list).await?;
```
Comment (waynexia) [inline src/storage/src/region/writer.rs:317]: Should this manifest update be done via `write_edit_and_apply()` method?

File: tests/cases/distributed/common
Comment (DevilExileSu) [reply tests/cases/distributed/common:1]: Sorry, this is accident.

File: src/storage/src/region/writer.rs (hunk 4 above)
Comment (DevilExileSu) [reply src/storage/src/region/writer.rs:317]: I don't think so, it doesn't modify meta and doesn't involve edit meta action. 
Maybe need to add `persist_manifest_version()` method.

File: src/storage/src/region/writer.rs [hunk 5]
```diff
@@ -354,6 +406,22 @@ impl<'a, S: LogStore> AlterContext<'a, S> {
     }
 }
 
+pub struct DropContext<'a, S: LogStore> {
+    pub shared: &'a SharedDataRef,
```
[the remark this answers is not in the document]
Comment (evenyag) [reply src/storage/src/region/writer.rs:410]: `SharedRegionDataRef` is so long : ( The `SharedDataRef` is under the `region` mod so it means `region::SharedDataRef`.

File: src/storage/src/engine.rs [hunk 6]
```diff
@@ -544,4 +553,99 @@ mod tests {
 
         assert!(engine.get_region(&ctx, "no such region").unwrap().is_none());
     }
+
+    #[tokio::test]
+    async fn test_drop_region() {
```
Comment (evenyag) [inline src/storage/src/engine.rs:558]: nitpick: `test_create_new_region()` also needs to create a region.  Could we wrap a helper function to create the engine and the region from specific temp dirs?

File: src/storage/src/engine.rs [hunk 7]
```diff
@@ -492,23 +498,35 @@ impl<S: LogStore> EngineInner<S> {
...
+    type TestEngine = EngineImpl<RaftEngineLogStore>;
+    type TestRegion = RegionImpl<RaftEngineLogStore>;
+
+    async fn create_engine_and_region(
+        tmp_dir: &TempDir,
+        region_name: &str,
+        region_id: u64,
+        ctx: &EngineContext,
+    ) -> (TestEngine, TestRegion) {
         let log_file_dir = create_temp_dir("test_engine_wal");
```
Comment (evenyag) [inline src/storage/src/engine.rs:525]: We should pass the log_file_dir as an argument so the temp dir won't be dropped after this function returns.

File: src/storage/src/region/writer.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (waynexia) [reply src/storage/src/region/writer.rs:410]: I have no idea what "data" it's when I'm looking at this, until I pull out my ide and jump to `SharedDataRef`

**Review by waynexia — APPROVED** [review]
LGTM, thanks

**Review by evenyag — APPROVED** [review]
LGTM
